src/main/java/com/google/firebase/remoteconfig/ConditionEvaluator.java

```diff
@@ -0,0 +1,266 @@
+    if (nestingLevel > MAX_CONDITION_RECURSION_DEPTH) {
+      logger.warn("Maximum condition recursion depth exceeded.");
+      return false;
+    }
+
+    if (condition.getOrCondition() != null) {
+      return evaluateOrCondition(condition.getOrCondition(), context, nestingLevel + 1);
+    } else if (condition.getAndCondition() != null) {
+      return evaluateAndCondition(condition.getAndCondition(), context, nestingLevel + 1);
+    } else if (condition.isTrue() != null) {
+      return true;
+    } else if (condition.isFalse() != null) {
+      return false;
+    } else if (condition.getCustomSignal() != null) {
+      return evaluateCustomSignalCondition(condition.getCustomSignal(), context);
+    }
+    logger.atWarn().log("Received invalid condition for evaluation.");
+    return false;
+  }
+
+
+  private boolean evaluateOrCondition(OrCondition condition, KeysAndValues context,
+      int nestingLevel) {
+    return condition.getConditions().stream()
+        .anyMatch(subCondition -> evaluateCondition(subCondition, context, nestingLevel + 1));
+  }
+
+  private boolean evaluateAndCondition(AndCondition condition, KeysAndValues context,
+      int nestingLevel) {
+    return condition.getConditions().stream()
+        .allMatch(subCondition -> evaluateCondition(subCondition, context, nestingLevel + 1));
+  }
+
+  private boolean evaluateCustomSignalCondition(CustomSignalCondition condition,
+      KeysAndValues context) {
+    CustomSignalOperator customSignalOperator = condition.getCustomSignalOperator();
+    String customSignalKey = condition.getCustomSignalKey();
+    ImmutableList<String> targetCustomSignalValues = ImmutableList.copyOf(
+        condition.getTargetCustomSignalValues());
+
+    if (targetCustomSignalValues.isEmpty()) {
+      logger.warn(String.format(
+          "Values must be assigned to all custom signal fields. Operator:%s, Key:%s, Values:%s",
+          customSignalOperator, customSignalKey, targetCustomSignalValues));
+      return false;
+    }
+
+    String customSignalValue = context.get(customSignalKey);
+    if (customSignalValue == null) {
+      return false;
+    }
+
+    switch (customSignalOperator) {
+      // String operations.
+      case STRING_CONTAINS:
+        return compareStrings(targetCustomSignalValues, customSignalValue,
+            (customSignal, targetSignal) -> customSignal.contains(targetSignal));
+      case STRING_DOES_NOT_CONTAIN:
+        return !compareStrings(targetCustomSignalValues, customSignalValue,
+            (customSignal, targetSignal) -> customSignal.contains(targetSignal));
+      case STRING_EXACTLY_MATCHES:
+        return compareStrings(targetCustomSignalValues, customSignalValue,
+            (customSignal, targetSignal) -> customSignal.equals(targetSignal));
+      case STRING_CONTAINS_REGEX:
+        return compareStrings(targetCustomSignalValues, customSignalValue,
+            (customSignal, targetSignal) -> Pattern.compile(targetSignal)
```

Comment:
Does it make sense to compile and cache this (if we expect this to be used more frequently)

Reply:
Should we also wrap this in a try-catch? Could this throw `java.util.regex.PatternSyntaxException`?